node/lib/packaging/LocalPackageSource.ts

```diff
@@ -0,0 +1,136 @@
+
+import { PackageInfo } from "./PackageInfo";
+import { PackageSource } from "./PackageSource";
+
+import * as fs from "mz/fs";
+import * as path from "path";
+
+/**
+ * Reads translator metadata and loads translators from a local directory
+ * where translator NPM packages have been installed.
+ */
+export class LocalPackageSource extends PackageSource {
+    /**
+     * Gets all subdirectories (not files) under a directory.
+     */
+    private static async getSubdirectoriesAsync(directoryPath: string): Promise<string[]> {
+        let directoryNames: string[] = [];
+        let childNames: string[] = await fs.readdir(directoryPath);
+
+        for (let i = 0; i < childNames.length; i++) {
+            if ((await fs.stat(path.join(directoryPath, childNames[i]))).isDirectory()) {
+                directoryNames.push(childNames[i]);
+            }
+        }
+
+        return directoryNames;
+    }
+
+    /**
+     * Path to the directory where packages are cached. In order to enable requiring
+     * modules in these packages without using relative paths, the path should be a
+     * node_modules directory that is included in the search path.
+     */
+    public readonly cacheDirectory: string;
+
+    /**
+     * Creates a new LocalPackageSource with a specified cache directory.
+     *
+     * @param {string} cacheDirectory  Path to the directory where packages are cached.
+     */
+    constructor(cacheDirectory: string) {
+        super();
+        this.cacheDirectory = cacheDirectory;
+    }
+
+    /**
+     * Gets information about all OpenT2T packages currently available from the source.
+     * CAUTION: This is a potentially slow operation.
```

Comment:
That may be possible, but I think that would be better for the application to handle that caching. In the case of a local package source, the cache could become out of date if the application downloads additional packages to the local cache. In the case of a remote package source, the cache could become out of date if new packages are published.

---
In reply to: [73290560](https://github.com/opent2t/opent2t/pull/2#discussion_r73290560) [](ancestors = 73290560)